plugin/backend/ipvariant.d

```diff
@@ -0,0 +1,906 @@
+        import std.exception : assumeUnique;
+
+        char[] buf;
+        buf.reserve(100);
+        toString((const(char)[] s) { buf ~= s; });
+        auto trustedUnique(T)(T t) @trusted {
+            return assumeUnique(t);
+        }
+
+        return trustedUnique(buf);
+    }
+}
+
+private:
+@safe:
+
+import cpptooling.data.representation : CppRoot, CppClass, CppMethod, CppCtor,
+    CppDtor, CFunction, CppNamespace;
+import cpptooling.data.type : LocationTag, Location;
+import cpptooling.data.symbol.container : Container;
+import dsrcgen.cpp : E;
+
+enum ClassType {
+    Normal,
+    Adapter,
+    Gmock
+}
+
+enum NamespaceType {
+    Normal,
+    TestDoubleSingleton,
+    TestDouble
+}
+
+/** Structurally transform the input to a stub implementation.
+ *
+ * TODO should handle StorageClass like cvariant do.
+ *
+ * Ignoring C functions and globals by ignoring the root ranges funcRange and
+ * globalRange.
+ *
+ * Params:
+ *  ctrl: removes according to directives via ctrl
+ */
+CppT rawFilter(CppT, LookupT)(CppT input, Controller ctrl, Products prod, LookupT lookup) @safe {
+    import std.array : array;
+    import std.algorithm : each, filter, map, filter;
+    import std.range : tee;
+    import application.types : FileName;
+    import cpptooling.data.representation : StorageClass, CppNs, CppNsStack;
+    import cpptooling.generator.utility : filterAnyLocation;
+    import cpptooling.utility : dedup;
+
+    // setup
+    static if (is(CppT == CppRoot)) {
+        auto filtered = CppRoot.make;
+    } else static if (is(CppT == CppNamespace)) {
+        auto filtered = input.dup; //CppNamespace.make(input.name);
+        assert(!input.isAnonymous);
+        assert(input.name.length > 0);
+    } else {
+        static assert("Type not supported: " ~ CppT.stringof);
+    }
+
+    // type specific
+    static if (is(CppT == CppNamespace)) {
+        // dfmt off
+        input.funcRange
+            .array()
+            .dedup
+            // by definition static functions can't be replaced by test doubles
+            .filter!(a => a.storageClass != StorageClass.Static)
+            // ask controller if the file should be mocked, and thus the node
+            .filterAnyLocation!(a => ctrl.doFile(a.location.file, cast(string) a.value.name ~ " " ~ a.location.toString))(lookup)
+            // pass on location as a product to be used to calculate #include
+            .tee!(a => prod.putLocation(FileName(a.location.file), LocationType.Leaf))
+            .each!(a => filtered.put(a.value));
+        // dfmt on
+
+    }
+
+    // Assuming that namespaces are never duplicated at this stage.
+    // The assumption comes from the structure of the clang AST.
+
+    // dfmt off
+    input.namespaceRange
+        .filter!(a => !a.isAnonymous)
+        .map!(a => rawFilter(a, ctrl, prod, lookup))
+        .each!(a => filtered.put(a));
+
+    input.classRange
+        .each!(a => filtered.put(a));
+
+    if (ctrl.doGoogleMock) {
```

Comment:
Google mock should be removed, I don't think it will have any use for this plugin (I guess).